### ISSUE TITLE
Screen is black with bloom and black artefact with DOF on a MacBook Pro 2017

### ISSUE THREAD

**Issue by CosmosTbear** [Unity-Technologies/PostProcessing#207]
Hi,
We have 2 issues with our game Kona on a MacBook Pro 2017
1. When using the bloom effect the screen is fully black
2. When we use the DOF, there is black artefacts
[https://youtu.be/hyTAktzOw_k](https://youtu.be/hyTAktzOw_k)

It may also happens on other mac machine but we can't be sure, we only have a Mac mini for our test on the mac version and everything works great.

We also use these effects on the stack:
Atialiasing
Eye Adaptation
Color Grading
Vignette
Dithering

Here are the machine spec:

13-inch model
2.3GHz dual-core Intel Core i5, Turbo Boost up to 3.6GHz, with 64MB of eDRAM
Configurable to 2.5GHz dual-core Intel Core i7, Turbo Boost up to 4.0GHz, with 64MB of eDRAM
Intel Iris Plus Graphics 640

**Comment by mcharbonneau** [comment]
Those black artefacts can also happen on Windows (we could only reproduce on a ATI card). There's a spot in our game where we have a decal using a shader from UBER. It seems to be linked to a factor of 2 we apply on normal mapping on this decal. It then introduces those artefacts when processing the depth of field. 

I would guess it creates values that the DOF is unable to process correctly in the depth texture, but it's hard to know what exactly.

![kona 2017-06-20 10-50-07-94](https://user-images.githubusercontent.com/6473235/27346729-0ebc8d0c-55bc-11e7-9bbb-c39f26739afd.jpg)

**Comment by Chman** [comment]
The black artefacts are NaN (Not-A-Number) or +/-Inf (Infinite) values. They generally happen when something goes wrong with the maths (like a division by zero). DoF only reads the input it's given, and invalid pixels will be blown out of proportion because it's trying to blur it (so it'll be spread out). So you'll want to check the shaders in passes before DoF (in @mcharbonneau I would look at the Decal shader).

The results can be different between GPUs because they don't all treat NaNs/Infs the same way (as there's no standard on this). Sometimes they output 0 (black pixel - so it doesn't hurt the following passes too much), sometimes they output a NaN or Inf (like in this case, doing maths with NaNs/Infs will break everything).

See my reply to issue #205 for more info about how to track down the shader responsible for these.

**Comment by Kink3d** [comment]
Related, @CosmosTbear does the Bloom black screen issue reproduce with only Bloom enabled? We have seen this sort of thing with low end mobile GPU but not on desktop yet.

**Comment by CosmosTbear** [comment]
Yeah only with the bloom effect enabled and when there seem to have a large amount of near white pixel on screen, when the rendered environement is darker the black screen is less frequent but get a blinking screen depending where you look.

[https://youtu.be/o0Bki5zQSes](https://youtu.be/o0Bki5zQSes)

**Comment by AlexandreFiset** [comment]
@Kink3d  Info has been added. This is really an annoying issue with dozens of people on Steam reporting problems on their Mac... Is there any known workaround for the Stack to work properly on newer Macs? We are going to buy a Macbook to better test things out. Maybe it is a Metal issue?

**Comment by EnricoMonese** [comment]
I've resolved a bunch of issues by switching to the new .NET version

**Comment by AlexandreFiset** [comment]
Any ETA on this? I mean, current V1 bloom and DOF aren't compatible with many Macs.

**Comment by CosmosTbear** [comment]
So... are you just ignoring the issue or...?

**Comment by Kink3d** [comment]
Hey,

We are looking in to this, but its proving difficult to pinpoint the exact cause. We will keep you posted.

**Comment by tatoforever** [comment]
I got the same issue on iOS but it's the other way around. If I turn on HDR i start to get black squares like the one in the picture on first post. On iOS GLES2.0/3.0 is not that bad, black squares are tiny. If I set it to Metal, the screen goes almost black with huge black squares everywhere. This is a very old issue since Unity added support to HDR on mobile. I don't think it's a driver issue, how other engines aren't having this same problem when they do HDR within their games? It is a Unity issue, pretty sure. When doing post processing like tonemapping and bloom. Setting the rendertexture to different precision values gives smaller or bigger black squares...

**Comment by tatoforever** [comment]
I'm aware that this is somehow related to the NaN issues everyone here is experiencing but none of my shaders are doing something wonky or weird and none of them should trigger NaN values because I'm the one writing them all. This issue come somewhere else, shader compiler, mesh import, etc? hard to know. :/
PS: Is there a way to get a frame snapshot on iOS devices to analyze on RenderDoc?

**Comment by Lestat** [comment]
My ZTE v7 has this issue, it seems it runs well in other devices (my users have reported).
